Unify outputs types

SarahAlidoost opened the pull request:
closes #640 
relates #691, #713, #711

See an overview of changes [here](https://github.com/dianna-ai/dianna/issues/640#issuecomment-1916752378).

cwmeijer commented on dianna/__init__.py:
```diff
@@ -71,7 +71,7 @@ def explain_image(model_or_function, input_image, method, labels, **kwargs):
 
     Returns:
-        One heatmap (2D array) per class.
+        An array (np.ndarray) containing the heat maps for each class.
```
Nice and thorough!
It would be even better if we could have type hints in the function signature.
We once decided (about a year ago) to start using type hints. We decided back then to not change the complete code at once, but only update code wherever we are working on it already. Now that you already created such thorough docstrings including input and output types, it shouldn't be such a big task to add those to the function signature as well:
def explain_image(model_or_function:Union(str, Callable), input_image:ndarray, method:string, labels:Iterable[int], **kwargs) -> ndarray:
If that's too much effort or is too troublesome in certain cases, only the output types would already help.
Let me know if you disagree to include this in this PR.

SarahAlidoost replied:
Good idea. I added types wherever possible. It would be nice if `mypy` is also added to the workflows to check the typing. see issue https://github.com/dianna-ai/dianna/issues/85

cwmeijer commented on dianna/methods/kernelshap_image.py:
```diff
@@ -100,7 +104,7 @@ def explain(
 
         Returns:
-            Explanation heatmap of shapley values for each class (np.ndarray).
+            Explanation heatmap for each class (np.ndarray).
```
I think I would prefer the old comment:
```suggestion
            Explanation heatmap of Shapley values for each class (np.ndarray).
```

SarahAlidoost commented:
> Thanks for your PR! Clearly, a lot of work went into it! I only reviewed 1 or 2 files yet. Could you have a look at my comment on type hints? I also see that the notebooks are failing. ![image](https://private-user-images.githubusercontent.com/6087314/308093601-b6220d25-2112-474d-a159-3cc1ec47beb5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNDU2NjIsIm5iZiI6MTcwOTA0NTM2MiwicGF0aCI6Ii82MDg3MzE0LzMwODA5MzYwMS1iNjIyMGQyNS0yMTEyLTQ3NGQtYTE1OS0zY2MxZWM0N2JlYjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjdUMTQ0OTIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDcxZjllY2MxYjY3ZTVkNjNiOWQxYzRhMGI5OGNhZmUyMmMxMzE2ZDAwY2Q1MGVlNGFkMGFmNGM4MWU5NDNlMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.vQbdrNSDecySgsqPjnEjce8vQKaYEdfxHCzwaCTdXiA) I think it's because you changed the output type of some explainer so it doesn't match the notebook's code anymore. Could you try and fix those?

Thanks for reviewing. I addressed your comment about typing. Regarding the notebook error, as discussed, they will be fixed  in a separate PR, see #711

cwmeijer commented on dianna/methods/kernelshap_image.py:
```diff
@@ -123,12 +127,27 @@ def explain(
+        for i, shap_value in enumerate(shap_values):
+            class_heat_map = heat_maps[i]
+            for index in self.image_segments.flat:
+                class_heat_map[self.image_segments == index] = shap_value[0][index - 1]
```
I'm a bit confused about what is in shap_values. It seems there is a list in there for every class. Then inside that list there is another level (hence '[0]') and only in there is the actual list of shap_values for that class. Except for that extra level, it makes sense to me, but maybe we can help the reader of the code a bit with more descriptive variable name?
Maybe shap_value_lists instead of shap_values to imply that there is a multilevel thing in there for instance?

SarahAlidoost replied:
Thanks, I agree. It is renamed now.

cwmeijer commented on dianna/methods/kernelshap_image.py:
```diff
@@ -123,12 +127,27 @@ def explain(
 
-        return shap_values, self.image_segments
+        # create heat_maps where shape is (n_classes, *image_segments.shape)
+        n_classes = len(shap_values)
+        heat_maps = np.zeros((n_classes, *self.image_segments.shape))
+
+        # fill the heat_maps with shap values for each class and segment
+        for i, shap_value in enumerate(shap_values):
+            class_heat_map = heat_maps[i]
+            for index in self.image_segments.flat:
+                class_heat_map[self.image_segments == index] = shap_value[0][index - 1]
+            heat_maps[i] = class_heat_map
```
You could consider extracting these lines into its own method and maybe even create a test for it.
Both are just suggestions, doing neither of these things at this point is totally fine with me.

SarahAlidoost replied:
I understand that the code will be more readable if this code block goes to its own function. However, I think it does not need to be a method of the class as the `heatmaps` or `image_segments` are already returned. Instead of a method, I refactored the code block as a function in this module.

cwmeijer reviewed (approved):
Thanks for this PR! Great addition to our code base and improvement in usability!
I made some suggestions, but all are to some extend details. Whether you change anything or not, I approve merging this PR.
🎖

elboyran commented:
Some checks should pass before we merge.

elboyran commented:
Yes, you are correct.

SarahAlidoost commented:
> Thanks for this PR! Great addition to our code base and improvement in usability! I made some suggestions, but all are to some extend details. Whether you change anything or not, I approve merging this PR. 🎖

Thank you for reviewing and nice comments :+1:

elboyran commented:
When do we merge this PR?
